freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/dns/DnsRecordService.java

```diff
@@ -55,23 +70,142 @@ private void deleteRegularRecords(FreeIpaClient freeIpaClient, DnsZone dnsZone,
+    }
+
+    @Retryable(value = RetryableFreeIpaClientException.class,
+            maxAttemptsExpression = RetryableFreeIpaClientException.MAX_RETRIES_EXPRESSION,
+            backoff = @Backoff(delayExpression = RetryableFreeIpaClientException.DELAY_EXPRESSION,
+                    multiplierExpression = RetryableFreeIpaClientException.MULTIPLIER_EXPRESSION))
+    @Measure(DnsRecordService.class)
+    public void addDnsARecord(String accountId, @Valid AddDnsARecordRequest request) throws FreeIpaClientException {
+        FreeIpaAndClient freeIpaAndClient = createFreeIpaAndClient(request.getEnvironmentCrn(), accountId);
+        LOGGER.info("Processing AddDnsARecordRequest: {}", request);
+        String zone = calculateZone(request.getDnsZone(), freeIpaAndClient);
+        Optional<DnsRecord> dnsRecord = ignoreNotFoundExceptionWithValue(() -> freeIpaAndClient.getClient().showDnsRecord(zone, request.getHostname()), null);
+        if (dnsRecord.isEmpty()) {
+            createDnsARecord(freeIpaAndClient.getClient(), zone, request.getHostname(), request.getIp(), request.isCreateReverse());
+        } else {
+            validateExistingARecordMatchesRequested(request.getIp(), dnsRecord.get());
+        }
+    }
+
+    private String calculateZone(String zoneFromRequest, FreeIpaAndClient freeIpaAndClient) throws FreeIpaClientException {
+        Optional<String> optionalZone = Optional.ofNullable(zoneFromRequest);
+        String zone = optionalZone.orElse(freeIpaAndClient.getFreeIpa().getDomain());
+        boolean providedZoneDifferentThanDefault = !freeIpaAndClient.getFreeIpa().getDomain().equals(StringUtils.removeEnd(zone, "."));
+        if (optionalZone.isPresent() && providedZoneDifferentThanDefault) {
+            validateZoneExists(optionalZone.get(), freeIpaAndClient.getClient());
+        }
+        return zone;
+    }
+
+    private void validateZoneExists(String zone, FreeIpaClient client) throws FreeIpaClientException {
+        Set<DnsZone> allDnsZone = client.findAllDnsZone();
+        String zoneInFreeIpaFormat = StringUtils.appendIfMissing(zone, ".");
+        boolean zoneMissing = allDnsZone.stream()
+                .map(DnsZone::getIdnsname)
+                .noneMatch(zoneFromIpa -> zoneFromIpa.equals(zoneInFreeIpaFormat));
+        if (zoneMissing) {
+            String msg = String.format("Zone [%s] doesn't exists", zone);
+            LOGGER.info(msg);
+            throw new BadRequestException(msg);
+        }
+    }
+
+    private void validateExistingARecordMatchesRequested(String ip, DnsRecord record) {
+        LOGGER.debug("Validating already existing record: {}", record);
+        if (!record.isARecord()) {
+            LOGGER.info("Record already exists and it's not an A record");
+            throw new DnsRecordConflictException("Record already exists and it's not an A record");
+        } else if (!record.getArecord().contains(ip)) {
+            LOGGER.info("Record already exists and the IP doesn't match");
```

Comment:
1 string would be enough instead of 2

Reply:
could you elaborate?

Reply:
You have the same text here and in next line:"Record already exists and the IP doesn't match". It would be more convenient to have only 1.

Reply:
on purpose. I didn't want to add multiple extra lines just to ensure the log and exception has the same message